nistats/glm.py

```diff
@@ -58,183 +47,394 @@ def data_scaling(Y):
-            the temporal variance model. Defaults to 'ar1'
-        bins : int, optional
-            Maximum number of discrete bins for the AR(1) coef histogram
-        """
-        if model not in ['ar1', 'ols']:
-            raise ValueError('Unknown model')
-
-        if Y.ndim == 1:
-            Y = Y[:, np.newaxis]
-
-        if Y.shape[0] != self.X.shape[0]:
-            raise ValueError('Response and predictors are inconsistent')
-
-        # fit the OLS model
-        ols_result = OLSModel(self.X).fit(Y)
-
-        # compute and discretize the AR1 coefs
-        ar1 = ((ols_result.resid[1:] * ols_result.resid[:-1]).sum(0) /
-               (ols_result.resid ** 2).sum(0))
-        ar1 = (ar1 * bins).astype(np.int) * 1. / bins
-
-        # Fit the AR model acccording to current AR(1) estimates
-        if model == 'ar1':
-            self.results_ = {}
-            self.labels_ = ar1
-            # fit the model
-            for val in np.unique(self.labels_):
-                m = ARModel(self.X, val)
-                self.results_[val] = m.fit(Y[:, self.labels_ == val])
+    # compute and discretize the AR1 coefs
+    ar1 = ((ols_result.resid[1:] * ols_result.resid[:-1]).sum(0) /
+           (ols_result.resid ** 2).sum(0))
+    ar1 = (ar1 * bins).astype(np.int) * 1. / bins
+
+    # Fit the AR model acccording to current AR(1) estimates
+    if noise_model == 'ar1':
+        results = {}
+        labels = ar1
+        # fit the model
+        for val in np.unique(labels):
+            model = ARModel(X, val)
+            results[val] = model.fit(Y[:, labels == val])
+    else:
+        labels = np.zeros(Y.shape[1])
+        results = {0.0: ols_result}
+    return labels, results
+
+
+def compute_contrast(labels, regression_result, con_val, contrast_type=None):
+    """ compute the specified contrast given an estimated glm
+
+    Parameters
+    ----------
+    labels : array of shape (n_voxels),
+        a map of values on voxels used to identify the correseponding the model
+
+    results : dictionary,
+        keys correspond to the differnt labels
+        values are RegressionResults instances corresponding to the voxels
+
+    con_val : numpy.ndarray of shape (p) or (q, p)
+        where q = number of contrast vectors and p = number of regressors
+
+    contrast_type : {None, 't', 'F' or 'tmin-conjunction'}, optional
+        type of the contrast.  If None, then defaults to 't' for 1D
+        `con_val` and 'F' for 2D `con_val`
+
+    Returns
+    -------
+    con : Contrast instance,
+        Yields the statistics of the contrast (effects, variance, p-values)
+    """
+    con_val = np.asarray(con_val)
+    dim = 1
+    if con_val.ndim > 1:
+        dim = con_val.shape[0]
+
+    if contrast_type is None:
+        if dim == 1:
+            contrast_type = 't'
         else:
-            self.labels_ = np.zeros(Y.shape[1])
-            self.results_ = {0.0: ols_result}
+            contrast_type = 'F'
```

Comment:
This is the kind of case where I find ternary statements to be slightly more readable. That may just be me though:

```python
contrast_type = 't' if dim == 1 else 'F'
```